### PR TITLE
EASY-1899 basic authentication failure

### PR DESCRIPTION
**Pull request by jo-pol**
Fixes EASY-1899 basic authentication fails registered/blocked users en invalid credentials

#### When applied it will
* no longer return a header `WWW-Authenticate`, thus a browser like chrome won't respond with a browser login pop-up
*  ~~additional servlet tests with registered and blocked users~~ -> new pull request
* 

#### How should this be manually tested?

* deploy this api version with ui [#120](https://github.com/DANS-KNAW/easy-deposit-ui/pull/120)
* login with proper and wrong users

#### Related pull requests on github
repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-                      | [PR#](PRlink)     | ..

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala [hunk 1]
```diff
@@ -16,31 +16,58 @@
...
                          password: String)
                         (implicit request: HttpServletRequest,
                          response: HttpServletResponse
                         ): Option[AuthUser] = {
-    authenticationProvider
-      .authenticate(userName, password)
-      .getOrElse(None) // TODO or .getOrElse(halt(ServiceUnavailable.logResponse))
-    // or BadGateWay / InternalServer error depending on the type of error ?
+    def haltWithInvalidUser = {
+      app halt Unauthorized(body = "invalid user-name/password combination").logResponse
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala]:
```suggestion
      app halt Unauthorized(body = "invalid username/password combination").logResponse
```

File: src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala [hunk 2]
```diff
@@ -16,31 +16,58 @@
...
+    }
+
+    def haltWithRegisteredUser = {
+      app halt Unauthorized(body = "please confirm your registration e-mail").logResponse
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala]:
```suggestion
      app halt Unauthorized(body = "please confirm your registration before logging in").logResponse
```

File: src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala [hunk 3]
```diff
@@ -16,31 +16,58 @@
...
+      app halt ServiceUnavailable(body = "login service temporarily not available").logResponse
+    }
+
+    authenticationProvider.authenticate(userName, password) match {
+      case Success(Some(AuthUser(id, UserState.active))) => Some(AuthUser(id, UserState.active))
+      case Success(Some(AuthUser(_, UserState.registered))) => haltWithRegisteredUser
+      case Success(None) |
+           Success(Some(AuthUser(_, UserState.blocked))) => haltWithInvalidUser
+      case Success(Some(AuthUser(id, state))) =>
+        logger.error(s"state [$state] not known for user with $id")
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala]:
```suggestion
        logger.error(s"unknown state '$state' for user '$id'")
```

[the remark this answers is not in the document]
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala]: No.

File: src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala [hunk 4]
```diff
@@ -16,31 +16,58 @@
...
+      case Success(Some(AuthUser(id, state))) =>
+        logger.error(s"unknown state '$state' for user '$id'")
+        haltWithInvalidUser
+      case Failure(e) =>
+        logger.error(e.getMessage, e)
+        haltWithFailure
+    }
   }
 
+
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala]:
```suggestion
```

**Comment by rvanheest** [comment]
Tested this in combination with https://github.com/DANS-KNAW/easy-deposit-ui/pull/120. Here are some test scenarios I ran:

* **login with correct username/password:** succeeds; proper things are logged by deposit-api; deposit-ui continues with logged-in user
* **try to login with incorrect username/password:** succeeds, deposit-api responds with a message that is displayed verbatim by deposit-ui (see image below); deposit-api logs the correct 401; deposit-api does not send a `WWW-Authenticate` challenge in the header.
  image
* **try to login with a valid username/password, but with a `REGISTERED` user state:** succeeds with https://github.com/DANS-KNAW-jp/easy-deposit-api/pull/7; logging is done correctly; perhaps the message should change a bit
  image
* **try to login with a valid username/password, but with a `BLOCKED` user state:** succeeds; logging is done correctly; perhaps add a `logger.warn` for this case.
  image

File: src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala [hunk 5]
```diff
@@ -16,31 +16,58 @@
...
-    authenticationProvider
-      .authenticate(userName, password)
-      .getOrElse(None) // TODO or .getOrElse(halt(ServiceUnavailable.logResponse))
-    // or BadGateWay / InternalServer error depending on the type of error ?
+    def haltWithInvalidUser = {
+      app halt Unauthorized(body = "invalid username/password").logResponse
+    }
+
+    def haltWithRegisteredUser = {
+      app halt Unauthorized(body = "please confirm your registration before logging in").logResponse
```
Comment (rvanheest) [inline src/main/scala/nl.knaw.dans.easy.deposit/authentication/EasyBasicAuthStrategy.scala]: See the comment with screenshots above, this message could be improved

```suggestion
      app halt Unauthorized(body = "please confirm your registration first").logResponse
```

**Comment by rvanheest** [comment]
@jo-pol as discussed: TODO tests for EasyBasicAuthStrategy.validate

**Pull request merged**